Create handlers for all endpoints

fzk273 opened the pull request:
Create adapters for datetime and duration
Create HttpTaskServer
Add toJson method to Task

Commit:
Create handlers for all endpoints
Create adapters for datetime and duration
Create HttpTaskServer
Add toJson method to Task

Ivanov-s-tmn commented on src/ru/prakticum/HttpTaskServer.java:
```diff
@@ -0,0 +1,21 @@
+public class HttpTaskServer {
+    private final Integer portNumber = 8080;
+
+    public void main() throws IOException {
+        HttpServer httpServer = HttpServer.create(new InetSocketAddress(portNumber), 0);
+        httpServer.createContext("/tasks", new TasksHandler());
+        httpServer.createContext("/subtasks", new SubtasksHandler());
+        httpServer.createContext("/epics", new EpicsHandler());
+        httpServer.createContext("/history", new HistoryHandler());
+        httpServer.createContext("/prioritized", new PriorityHandler());
+        httpServer.start();
+    }
```
Создание и инициализацию сервера делай в конструкторе, добавь методы старт и стоп, через них ты уже сможешь запускать `HttpTaskServer` и останавливать его откуда угодно. Тут метод main не нужен.

Добавь возможность работы с разными реализациями таск менеджера и передавай его в конструктор при создании `HttpTaskServer`

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/Main.java:
```diff
@@ -9,20 +9,30 @@
 
 //        inMemoryTaskManagerUserStory(nowTime, oneHour);
-        fileBackedTaskManagerUserStory(nowTime, oneHour);
+//        fileBackedTaskManagerUserStory(nowTime, oneHour);
+        //TODO не уверен, что правильно понял из тз, как запускать сервер. нид хэлп
+
+        HttpTaskServer httpTaskServer = new HttpTaskServer();
+        try {
+            httpTaskServer.main();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
Ты просто создаешь его через конструктор и передаешь нужный менеджер `HttpTaskServer taskServer = new HttpTaskServer(manager);`
потом уже можешь запустить его `taskServer.start()` и остановить когда не нужен `taskServer.stop()`

fzk273 replied:
поправил, спасибо за разъяснение

Ivanov-s-tmn commented on src/ru/prakticum/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,51 @@
+public class BaseHttpHandler {
+    protected void sendText(HttpExchange h, String text) throws IOException {
+        Gson jsonBuilder = new GsonBuilder().create();
+        String jsonResponce = jsonBuilder.toJson(text);
+        byte[] resp = jsonResponce.getBytes(StandardCharsets.UTF_8);
+        h.getResponseHeaders().add("Content-Type", "application/json;charset=utf-8");
+        h.sendResponseHeaders(200, resp.length);
+        h.getResponseBody().write(resp);
+        h.close();
+    }
+
+    protected void sendCreated(HttpExchange h, String text) throws IOException {
+        Gson jsonBuilder = new GsonBuilder().create();
+        String jsonResponce = jsonBuilder.toJson(text);
+        byte[] resp = jsonResponce.getBytes(StandardCharsets.UTF_8);
+        h.getResponseHeaders().add("Content-Type", "application/json;charset=utf-8");
+        h.sendResponseHeaders(201, resp.length);
+        h.getResponseBody().write(resp);
+        h.close();
+    }
+
+    public void sendNotFound(HttpExchange h, String text) throws IOException {
+        Gson jsonBuilder = new GsonBuilder().create();
+        String jsonResponce = jsonBuilder.toJson(text);
+        byte[] resp = jsonResponce.getBytes(StandardCharsets.UTF_8);
+        h.getResponseHeaders().add("Content-Type", "application/json;charset=utf-8");
+        h.sendResponseHeaders(400, resp.length);
+        h.getResponseBody().write(resp);
+        h.close();
+    }
+
+    public void sendHasInteractions(HttpExchange h, String text) throws IOException {
+        Gson jsonBuilder = new GsonBuilder().create();
+        String jsonResponce = jsonBuilder.toJson(text);
+        byte[] resp = jsonResponce.getBytes(StandardCharsets.UTF_8);
+        h.getResponseHeaders().add("Content-Type", "application/json;charset=utf-8");
+        h.sendResponseHeaders(406, resp.length);
+        h.getResponseBody().write(resp);
+        h.close();
+    }
```
Методы дублируются, отличаются только статусы. Вообще, достаточно одного одного метода, `sendText()` для успешных запросов, можно передавать в него статус ответа. А для запросов с ошибкой достаточно возвращать ошибку в хедере и все
```java
h.sendResponseHeaders(406, resp.length);
h.close();
```

fzk273 replied:
оставил коммент в тудушке. продублирую ещё сюда: "если в случае ошибки отправлять без боди (как в комменте на пр) тогда запрос не возвращается корректно"

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,74 @@
+import java.time.LocalDateTime;
+import java.util.stream.Collectors;
+
+//TODO практически уверен, что копипастить хендлер не лучшая идея. Как реализовать лучше?
```
Это не совсем копипаста, все-таки у них разная логика, они переопределяют метод `handle()` со своей логикой, можно вынести только отдельные элементы, по типу инициализации gson

fzk273 replied:
спасибо! поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,74 @@
+        TaskManager taskManager = Managers.getFileBackedTaskManager();
+        switch (method) {
+            case "POST":
+                Gson gson = new GsonBuilder()
+                        .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeAdapter())
+                        .registerTypeAdapter(Duration.class, new DurationAdapter()).create();
```
gson проинициализируй в конструкторе, можно вынести в базовый хендлер

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,74 @@
+                        .registerTypeAdapter(Duration.class, new DurationAdapter()).create();
+                Epic epic = gson.fromJson(body, Epic.class);
+                Epic newEpic = taskManager.createEpic(epic);
+                super.sendCreated(httpExchange, newEpic.toJson());
```
через супер вызывать не надо, они наследуются

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,74 @@
+            case "GET":
+                if (splittedPath.length > 3 && splittedPath[3].equals("subtasks")) {
+                    Integer id = Integer.parseInt(splittedPath[2]);
+                    String epicSubtasks = taskManager.getEpicSubtasks(taskManager.getEpicById(id)).stream()
```
Преобразуй через `gson.toJson()`

fzk273 replied:
так сильно проще и json не кривой получается. Спасибо поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,74 @@
+                        super.sendNotFound(httpExchange, "there is no subtask with id: " + id);
+                    }
+                } else {
+                    String epicsJson = taskManager.getEpics().stream()
+                            .map(Epic::toJson)
+                            .collect(Collectors.joining(","));
+                    super.sendText(httpExchange, "{" + epicsJson + "}");
```
Тут тоже, все преобразование через gson

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/HistoryHandler.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.IOException;
+import java.util.stream.Collectors;
+
+//TODO не понимаю как забирать один и тот же инстанс TaskManager во всех хендлерах, поэтому история невалидна
+// нужна помощь.
```
Оставил об это комментарий в `HttpTaskServer` при инициализации хендлеров в конструкторе сервера нужно передавать один и тот же таск менеджер

fzk273 replied:
спасибо за разъяснения. я чёт тупанул.
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/HistoryHandler.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void handle(HttpExchange httpExchange) throws IOException {
+        String method = httpExchange.getRequestMethod();
+        TaskManager taskManager = Managers.getFileBackedTaskManager();
```
А почему файловый менеджер используешь

fzk273 replied:
убрал, теперь передаю в конструкторе

Ivanov-s-tmn commented on src/ru/prakticum/handlers/SubtasksHandler.java:
```diff
@@ -0,0 +1,75 @@
+                SubTask subTask = gson.fromJson(body, SubTask.class);
+
+                if (subTask.getId() != null) {
+                    //TODO вообще не понимаю как реализоывать проверку на пересечение из этого класа. need help
+//                    super.sendHasInteractions(httpExchange, "subtask is overlapping");
```
У тебя проверка на пересечения вшита в логику методов таск менеджера, зачем это делать тут?

fzk273 replied:
это было в тексте задания. без этого я не понимаю как понять что таска не прошла проверку и кинуть 406

Ivanov-s-tmn replied:
Выбрасывать исключение из таск менеджера при пересечении и отлавливать его здесь

fzk273 replied:
сделал

Ivanov-s-tmn commented on src/ru/prakticum/tasks/Task.java:
```diff
@@ -86,6 +94,14 @@ public String toString() {
                 + ",";
     }
 
+    public String toJson() {
+        Gson gson = new GsonBuilder()
+                .registerTypeAdapter(Duration.class, new DurationAdapter())
+                .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeAdapter())
+                .create();
+        return gson.toJson(this);
+    }
```
Просто вызывать `gson.toGson()` при формировании ответа в хендлерах

fzk273 replied:
поправил

Commit:
refactor handlers
add tests to all classes

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -1,73 +1,59 @@
-import java.util.stream.Collectors;
 
-//TODO практически уверен, что копипастить хендлер не лучшая идея. Как реализовать лучше?
 public class EpicsHandler extends BaseHttpHandler implements HttpHandler {
```
Можно имплементировать интерфейс в `BaseHttpHandler`, тут уже это будет не нужно

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -1,73 +1,59 @@
+                    sendText(httpExchange, epicSubtasks, 200);
                 } else if (splittedPath.length > 2) {
                     Integer id = Integer.parseInt(splittedPath[2]);
                     Epic epic1 = taskManager.getEpicById(id);
```
Старайся не использовать переменные с индексами 1, 2 и тд. Это не делает код понятнее)

fzk273 replied:
я отрефакторил, но как мне кажется сильно красивее не получилось

Ivanov-s-tmn commented on test/ru/prakticum/http/SubTasksHandlerTest.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Test
+    public void subTaskOverlappingIs406() {
+        //TODO не понимаю как имлементировать, поскольку не получаю 406 в создании таски
```
Оставил ответ на комментарий

fzk273 replied:
увидел, поправил! спасибо!

Ivanov-s-tmn commented on test/ru/prakticum/http/EpicsHandlerTest.java:
```diff
@@ -0,0 +1,104 @@
+public class EpicsHandlerTest extends InitHandlers {
+    private String epicJson;
+
+    @BeforeAll
+    public static void initServer() throws IOException {
+        InitHandlers.initServer();
+    }
+
+    @BeforeEach
+    public void init() {
+        super.init();
+
+    }
+
+    @AfterAll
+    public static void terminate() {
+        InitHandlers.terminate();
+    }
+
```
Эти методы наследуются, их не нужно переопределять каждый раз

fzk273 replied:
убрал.

Ivanov-s-tmn commented on test/ru/prakticum/http/EpicsHandlerTest.java:
```diff
@@ -0,0 +1,104 @@
+                .POST(HttpRequest.BodyPublishers.ofString(epicJson, StandardCharsets.UTF_8))
+                .build();
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+        Assertions.assertEquals(201, response.statusCode());
```
Недостаточно проверить только статус запроса, нужно проверить еще сам результат, после добавления сходить в таск менеджер и проверить, что задача добавилась. 
Это касается всех тестов

fzk273 replied:
добавил базовые проверки  и переписал чутка тесты

Ivanov-s-tmn commented on test/ru/prakticum/http/EpicsHandlerTest.java:
```diff
@@ -0,0 +1,104 @@
+        request = HttpRequest.newBuilder(uri.resolve("./epics/" + epicId)).GET().build();
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+        Assertions.assertEquals(200, response.statusCode());
+        //TODO тут постоянно возникает java.io.IOException: HTTP/1.1 header parser received no bytes не понимаю, что не так
+        // если только этот тест отдельно запустить, всё норм
```
Сложно сказать, нужно продебажить и посмотреть что происходит на этапе ошибки, так будет понятно.
Возможно перемудрил с настройками тестов в плане инициализации серверов, клиентов и тд.

fzk273 replied:
да, я и правда перемудрил, но в итоге удалось задебажить и устаканить тесты

Commit:
refactor tests
add implementation to BaseHttpHandler
refactor get method in epic handler
refactor post in SubtasksHandler
refactor post in TasksHandler

Commit:
fix linting

Commit:
fix linting

Ivanov-s-tmn commented on src/ru/prakticum/handlers/BaseHttpHandler.java:
```diff
@@ -35,4 +36,7 @@ protected void sendText(HttpExchange h, String text, Integer responseCode) throw
         h.close();
     }
 
+    @Override
+    public void handle(HttpExchange httpExchange) throws IOException {
+    }
```
Сделай класс абстрактным, тогда не нужно будет объявлять пустой метод)

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -57,4 +41,35 @@ public void handle(HttpExchange httpExchange) throws IOException {
+                sendText(httpExchange, "there is no epic with id: " + epicId, 404);
+            }
+        } else if (splitPath.length == 4 && isSubtasksInPath) {
+            if (taskManager.getEpicById(epicId) != null) {
```
Смотри, ты делаешь тут лишний вызов метода таск менеджера. Тебе нужно вернуть список сабтасок эпика, но ты еще дополнительно вызываешь метод получения по айди, который в свою очередь добавляет в историю, хотя логически это некорректно.

Не дергай лишний раз ненужные методы

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/SubtasksHandler.java:
```diff
@@ -24,14 +23,17 @@ public void handle(HttpExchange httpExchange) throws IOException {
+                        taskManager.updateSubtask(subTask);
+                        sendText(httpExchange, gson.toJson(subTask), 201);
+                    }
+                } catch (IllegalArgumentException exception) {
```
Для исключений, которые ты сам пробрасываешь лучше делать отдельные свои типы, как это мы делали с файловым менеджером при ошибках записи/чтения. Так ты сможешь проще управлять их обработкой.  `IllegalArgumentException` на самом деле может выскочит в большом количестве случаев

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/SubtasksHandler.java:
```diff
@@ -24,14 +23,17 @@ public void handle(HttpExchange httpExchange) throws IOException {
-                    sendText(httpExchange, gson.toJson(subTask), 201);
+                try {
+                    if (subTask.getId() == null) {
+                        SubTask subTask1 = taskManager.createSubtask(subTask);
```
В названиях переменных старайся избегать индексов

fzk273 replied:
поправил

Ivanov-s-tmn commented on test/ru/prakticum/http/EpicsHandlerTest.java:
```diff
@@ -57,30 +40,22 @@ public void epicCreationIs201() throws IOException, InterruptedException {
-                .build();
-        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
-        Assertions.assertEquals(201, response.statusCode());
+        Assertions.assertEquals(1, taskManager.getEpics().size());
```
Размер списка тоже достаточно скудная проверка) 

Лучше все-таки запросить у менеджера созданную задачу по айди и сравнить поля

fzk273 replied:
поправил

Commit:
refactor tests
Add new exception for overlapping tasks
refactor BaseHttpHandler to abstract class
rename variables to variables without indexes
remove unnesesary calls of taskManager

Ivanov-s-tmn commented on src/ru/prakticum/exceptions/TaskOverlappingException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.prakticum.exceptions;
+
+public class TaskOverlappingException extends IllegalArgumentException {
```
Свои исключения лучше всегда наследуй от непроверяемого `RuntimeException`, так ты не будешь обязывать себя их обрабатывать

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -56,18 +56,18 @@ public void getMethod(HttpExchange httpExchange, String[] splitPath) throws IOEx
         } else if (splitPath.length == 4 && isSubtasksInPath) {
-            if (taskManager.getEpicById(epicId) != null) {
+            try {
                 String epicSubtasks = gson.toJson(taskManager.getEpicSubtasks(taskManager.getEpicById(epicId)));
```
Тут у тебя тоже вызывается метод `taskManager.getEpicById(epicId)` внутри `getEpicSubtasks()`, просто достань айди эпика из запроса и вызови `taskManager.getEpicSubtasks(epicId)`
Метод `getEpicSubtasks` вообще должен принимать айди, а не сам эпик, это было в ТЗ 4 спринта, упустил этот момент

fzk273 replied:
пофиксил

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -56,18 +56,18 @@ public void getMethod(HttpExchange httpExchange, String[] splitPath) throws IOEx
                 sendText(httpExchange, epicSubtasks, 200);
-            } else {
-                sendText(httpExchange, "there is epic with id: " + epicId, 404);
+            } catch (NullPointerException e) {
```
Обрабатывать NPE плохая практика, его не нужно допускать в целом в коде.

fzk273 replied:
поправил, но решение не оптимальное

Commit:
refactor getEpicById
change exception parent to RuntimeException
refactor get method in epic handler

Ivanov-s-tmn commented on src/ru/prakticum/handlers/EpicsHandler.java:
```diff
@@ -63,10 +65,11 @@ public void getMethod(HttpExchange httpExchange, String[] splitPath) throws IOEx
-            try {
-                String epicSubtasks = gson.toJson(taskManager.getEpicSubtasks(taskManager.getEpicById(epicId)));
+            boolean epicExists = taskManager.getEpics().stream().anyMatch(epic -> epic.getId().equals(epicId));
+            if (epicExists) {
+                String epicSubtasks = gson.toJson(taskManager.getEpicSubtasks(epicId));
                 sendText(httpExchange, epicSubtasks, 200);
-            } catch (NullPointerException e) {
+            } else {
                 sendText(httpExchange, "there is no epic with id: " + epicId, 404);
             }
```
Тебе в целом достаточно тут сделать один вызов и больше ничего.
`String epicSubtasks = gson.toJson(taskManager.getEpicSubtasks(epicId));`
В итоге в ответ всегда вернется 200 статус со списком сабтасок или с пустым списком, если эпик пустой или эпика не существует. Тут тебе нужно именно логику самого метода `getEpicSubtasks` под это подстроить.
Ты передаешь в него `epicId`, первое, что тебе нужно сделать - проверить на наличие эпика в хешмапе, это уже избавит тебя от возможного возникновения NPE. Потом ты уже работаешь с этим эпиков, если он не null и больше никаких проблем в не будет)
